Syntax sugar: class for iterating tokens in c++11 style

umanamente opened the pull request:
1. Added `IterateTokens` class which can be used to iterate tokens as C++11 range-based loops.

```cpp
for (const auto &tok : IterateTokens(scope)) {
  // ...
}
```

2. Added macro `ITERATE_TOKENS(tok, ...)`, which generates regular for-begin-end loop:
```cpp
for (ITERATE_TOKENS(tok, start, end)) { ... }
for (ITERATE_TOKENS(tok, scope)) { ... }
for (ITERATE_TOKENS(tok, mTokenizer)) { ... }
```

There was already `tokenrange.h`, which contained an iterator draft, but it wasn't used anywhere, and also it didn't support token assignment inside loops, so I removed it and replaced with `tokeniterators.h`.

# Usage:

## Start to end

#### Before:
```cpp
for (const Token *tok = start; tok && tok != end; tok = tok->next()) {
```

#### After:
```cpp
for (const auto &tok : IterateTokens(start, end)) {
```
or
```cpp
for (ITERATE_TOKENS(tok, start, end)) {
```


## Scopes
#### Before:
```cpp
for (const Token* tok = scope->bodyStart->next(); tok != scope->bodyEnd; tok = tok->next()) {
```

#### After:
```cpp
for (const auto &tok : IterateTokens(scope)) {
```
**OR**
```cpp
for (const auto &tok : IterateTokens::ScopeIncludeBraces(f.functionScope))
```
**OR**
```cpp
for (ITERATE_TOKENS(tok, scope)) { ... }
```


## Tokenizer
#### Before:
```cpp
for (const Token *tok = mTokenizer->tokens(); tok; tok = tok->next()) {
```

#### After:
```cpp
for (const auto &tok : IterateTokens(mTokenizer)) {
```
**OR**
```cpp
for (ITERATE_TOKENS(tok, mTokenizer)) { ... }
```


## IterateTokens::UntilLinked

#### Before:
```cpp
for (const Token *operTok = operStart; operTok != operStart->link(); operTok = operTok->next()) {
```

#### After:
```cpp
for (const auto &tok : IterateTokens::UntilLinked(operStart)) {
```

pfultz2 commented on lib/tokeniterators.h:
```diff
@@ -0,0 +1,141 @@
+ */
+class IterateTokens {
+
+protected: // forbid to create iterator instances by user to avoid unexpected copy behavior
```
Thats not a valid iterator, c++ iterators are required to be copy_constructible, copy_assignable, and swappable.

pfultz2 commented on lib/tokeniterators.h:
```diff
@@ -0,0 +1,141 @@
+    class iterator {
+    public:
+        explicit iterator(const Token* tok)
+            : m_sharedTokHolder(std::make_shared<TokenHolder>(tok))
```
I dont think dynamic allocation is good idea to do when constructing an iterator.

pfultz2 replied:
The `TokenHolder` can be stored in the `IterateTokens` class instead of using dynamic memory.

pfultz2 commented on lib/tokeniterators.h:
```diff
@@ -0,0 +1,141 @@
+            return *this;
+        };
+
+        iterator& operator++() {
```
You are missing the post-increment.

pfultz2 replied:
Iterators require post increment so we can't leave it unimplemented.

umanamente replied:
Range-based loops don't require post-increment, and I don't think it will be used in other places. Could we keep it simple?

umanamente commented on lib/tokeniterators.h:
```diff
@@ -0,0 +1,141 @@
+            return m_sharedTokHolder->m_tok;
+        };
+        
+        const iterator& operator*() const {
```
It is meant to be compatible with existing code, that's why this hack is required (to catch assignment operations).

There is a conversion operator which makes it possible to be used as function argument where `const Token*` is expected, and `->` operator.

umanamente commented on lib/tokeniterators.h:
```diff
@@ -0,0 +1,141 @@
+protected: // forbid to create iterator instances by user to avoid unexpected copy behavior
+
+    class iterator {
+    public:
```
Same as above, the goal wasn't to provide a 100% valid iterator, it only has to provide bare minimum for range-based loops to work. It wasn't meant to be used in standard library functions, hence I don't think that iterator specifications are necessary here.

umanamente replied:
Added iterator traits

danmar commented:
> Handles "jumping" correctly (assignment of token inside loop body):

This looks weird to me. It's not consistent with normal range for loop behavior. So I wonder if this handling can be removed.

pfultz2 commented:
I agree it's not consistent and I don't think it would be obvious to most developers what is being done.

umanamente commented:
Removed that feature (made assignment forbidden) and added `ITERATE_TOKENS(tok, ...)` macro which expands into regular for-begin-end loop and could replace existing loops without any impact.

jpyllman commented:
I think this way is very good. Cause there are many different ways this is written in the code. And it actually introduces possible 'buffer' overflows. I just wonder if the check for 'end of range' should be like in precedes() defined in lib/astutils.h.

Also a small note, could we make precedes() inline? It actually make a noticeable difference. And if we start using this c++11-ish style of for loop it should be inlined.
